src/catsRouter.js

```diff
@@ -1,164 +1,169 @@
-    const isUpdated = CatRepository.updateCatById(req.id, req.object);             // [ {}, {}, {} ]
-    if (isUpdated) {
-        console.log(`cat id '${req.id}' successfully updated`)
-        res.send(isUpdated); // may need to use: res.send(JSON.stringify(updatedCat)); 
-    } else {
-        const newError = new Error(`cat id '${req.id}' not found in database`);
-        newError.status = 404; 
-        next(newError); 
+    // checks object data format
+    checkObjFormat (req, res, next) { 
+        const objectConstructor = ({}).constructor; // in strict mode, 'this' in functions is undefined. 
+        if (req.body.constructor === objectConstructor) { // within methods, 'this' refers to the object the method is owned by. 
+            console.log("verified request data is an object") // this does not apply to nested functions, where you declare a function within an object's method
+            next();
+        } else {
+            const newError = new Error ('Request data is not an object - check format');
+            newError.status = 400;
+            return next(newError);
+        }
     }
-});
 
-
-// DEL route
-catsRouter.delete('/:id', isIdNum, (req, res, next) => {
-    const isDeleted = CatRepository.deleteCatById(req.id); 
-    if (isDeleted) {
-        res.status(204).send();
-    } else {
-        const newError =  new Error(`cat id '${req.id}' not found in database`)
-        newError.status = 404;
-        return next(newError)
+    // checks object keys  
+    checkObjKeys (req, res, next) {  
+        
+        const validKeys = ["name", "ageInYears", "favouriteToy", "description", "breedId"];
+        
+        function keyNotPresent (key) { 
+            console.log(`is ${key} present?`)
+            return !validKeys.includes(key) // returns true if key is NOT one of the possibleKeys
+        }
+        
+        const objToCheck = req.body;                
+        console.log("Checking object: " + objToCheck);
+        const arrayOfKeys = Object.keys(objToCheck);      
+        console.log("The keys to check: " + arrayOfKeys);
+        
+        if (arrayOfKeys.length < 1) {
+            const newError = new Error('no parameters provided');
+            newError.status = 404;
+            return next(newError);
+        }
+        
+        let invalidKey = arrayOfKeys.filter(keyNotPresent); // returns an array of elements that meet the criteria 
+        
+        if (invalidKey[0]) { 
+            console.log(`Property '${invalidKey[0]}' is invalid`);
+            const newError = new Error(`Property '${invalidKey[0]}' is invalid`);
+            newError.status = 400;
+            return next(newError);
+        } else {
+            console.log('object keys checked'); 
+            req.object = objToCheck; 
+            next();
+        }
+    } // FIXME this seems convoluted just to check if an object key is valid 
```

Comment:
There are probably a few things which could be altered here

Reply:
Well duh

Reply:
Excue my flippancy :) Let me know if I got all the 'things' in my latest PR.